File: server/src/main/java/com/bomberman/utils/Collisions.java
package com.bomberman.utils;

import com.bomberman.constants.Configuration;
import com.bomberman.entities.Position;
import com.bomberman.entities.mapobject.MapObject;
import com.bomberman.game.GameState;
import java.util.List;
import java.util.Objects;

public class Collisions {

  public static boolean isInExplosionRange(
      final int range,
      final MapObject fromObject,
      final MapObject toObject,
      final GameState gameState) {

    if (isInExplosionRange(range, fromObject, toObject, gameState, true, true)) {
      return true;
    }
    if (isInExplosionRange(range, fromObject, toObject, gameState, false, true)) {
      return true;
    }
    if (isInExplosionRange(range, fromObject, toObject, gameState, true, false)) {
      return true;
    }
    return isInExplosionRange(range, fromObject, toObject, gameState, false, false);
  }

  private static boolean isInExplosionRange(
      final int range,
      final MapObject fromObject,
      final MapObject toObject,
      final GameState gameState,
      final boolean isDirectionDown,
      final boolean xAxis) {
    for (int i = 1; i <= range; i++) {
      final float xPos;
      final float yPos;
      if (xAxis) {
        yPos = fromObject.getPosition().getY();
        if (isDirectionDown) {
          xPos = fromObject.getPosition().getX() - (i * fromObject.getTextureWidth());
        } else {
          xPos = fromObject.getPosition().getX() + (i * fromObject.getTextureWidth());
        }
      } else {
        xPos = fromObject.getPosition().getX();
        if (isDirectionDown) {
          yPos = fromObject.getPosition().getY() - (i * fromObject.getTextureHeight());
        } else {
          yPos = fromObject.getPosition().getY() + (i * fromObject.getTextureHeight());
        }
      }

      final var position = new Position(xPos, yPos);

      final boolean willHitWall =
          gameState.getWalls().stream()
              .anyMatch(wall -> Collisions.isAtSameSquare(wall.getPosition(), position));
      if (willHitWall) {
        break;
      }

      if (isAtSameSquare(position, toObject.getPosition())) {
        return true;
      }
    }
    return false;
  }

  public static boolean isOutOfBound(final MapObject mapObject, final Position newPosition) {
    return newPosition.getX() + mapObject.getTextureWidth() > Configuration.MAP_WIDTH
        || newPosition.getY() + mapObject.getTextureHeight() > Configuration.MAP_HEIGHT;
  }

  public static boolean willCollide(
      final MapObject mapObject, final Position newPosition, final List<MapObject> obstacles) {
    return obstacles.stream().anyMatch($obstacle -> willCollide(mapObject, newPosition, $obstacle));
  }

  public static boolean willCollide(
      final MapObject movableObject, final Position newPosition, final MapObject obstacle) {

    if (Objects.equals(newPosition, obstacle.getPosition())) {
      return true;
    }

    final Position l1 = newPosition.getTopLeft(movableObject);
    final Position r1 = newPosition.getBotRight(movableObject);
    final Position l2 = obstacle.getTopLeft();
    final Position r2 = obstacle.getBotRight();

    // If one rectangle is on left side of other
    if (l1.getX() > r2.getX() || l2.getX() > r1.getX()) {
      return false;
    }

    // If one rectangle is above other
    if (l1.getY() < r2.getY() || l2.getY() < r1.getY()) {
      return false;
    }

    return true;
  }

  public static boolean isAtSameSquare(
      final Position firstPosition, final Position secondPosition) {
    return Objects.equals(firstPosition, secondPosition);
  }
}

File: server/src/main/kotlin/com/bomberman/game/Collisions.kt
package com.bomberman.game

import com.bomberman.MAP_HEIGHT
import com.bomberman.MAP_WIDTH
import com.bomberman.entity.Position
import com.bomberman.entity.mapobject.MapObject
import com.bomberman.entity.mapobject.movable.Movable
import com.bomberman.entity.mapobject.pickup.Pickup

fun MapObject.isInExplosionRange(range: Int, toObject: MapObject,
    collideObjects: List<MapObject>): Boolean =
    when {
      this.isInExplosionRange(range, toObject, collideObjects, true, true) -> {
        true
      }
      this.isInExplosionRange(range, toObject, collideObjects, false, true) -> {
        true
      }
      this.isInExplosionRange(range, toObject, collideObjects, true, false) -> {
        true
      }
      this.isInExplosionRange(range, toObject, collideObjects, false, false) -> {
        true
      }
      else -> {
        false
      }
    }

fun MapObject.isInExplosionRange(
    range: Int,
    toObject: MapObject,
    collideObjects: List<MapObject>,
    isDirectionDown: Boolean,
    xAxis: Boolean): Boolean {

  for (i in 1..range) {
    val xPos: Float
    val yPos: Float
    if (xAxis) {
      yPos = this.position.y
      xPos = if (isDirectionDown) {
        this.position.x - i * this.textureWidth
      } else {
        this.position.x + i * this.textureWidth
      }
    } else {
      xPos = this.position.x
      yPos = if (isDirectionDown) {
        this.position.y - i * this.textureHeight
      } else {
        this.position.y + i * this.textureHeight
      }
    }
    val position = Position(xPos, yPos)
    val willHitCollideObject = collideObjects.any {
      it.isAtSameSquare(position)
    }
    if (willHitCollideObject) {
      break
    }
    if (toObject.isAtSameSquare(position)) {
      return true
    }
  }
  return false
}

fun MapObject.isOutOfBound(newPosition: Position): Boolean =
    newPosition.x + this.textureWidth > MAP_WIDTH
        || newPosition.y + this.textureHeight > MAP_HEIGHT
        || newPosition.x < 0
        || newPosition.y < 0

fun MapObject.willCollide(newPosition: Position, obstacles: List<MapObject>): Boolean =
    obstacles.any { this.willCollide(newPosition, it) }

fun <T> T.firstPickupCollision(newPosition: Position,
    pickups: List<Pickup>): Pickup? where T : Movable, T : MapObject =
    pickups.firstOrNull { this.willCollide(newPosition, it) }

fun MapObject.willCollide(
    newPosition: Position, obstacle: MapObject): Boolean {
  if (newPosition == obstacle.position) {
    return true
  }
  val l1 = newPosition.topLeft(this)
  val r1 = newPosition.botRight(this)
  val l2 = obstacle.topLeft
  val r2 = obstacle.botRight
  // If one rectangle is on left side of other
  if (l1.x > r2.x || l2.x > r1.x) {
    return false
  }
  // If one rectangle is above other
  return when {
    l1.y < r2.y || l2.y < r1.y -> {
      false
    }
    else -> true
  }
}

fun MapObject.isAtSameSquare(secondPosition: Position): Boolean =
    this.position == secondPosition